File: 09wp-master/src/utils/coverProxy.ts
/** 资源是否应按「TG 图片返代」处理（与后台 source / external_id / 封面域一致） */
export function shouldApplyTgCoverProxy(
  resource: Record<string, unknown> | null | undefined,
  coverRaw: string,
): boolean {
  if (!resource) return false
  const src = String(resource.source ?? '').trim().toLowerCase()
  if (src === 'telegram') return true
  const ext = String(resource.external_id ?? '').trim().toLowerCase()
  if (ext.startsWith('tg:')) return true
  const u = coverRaw.trim().toLowerCase()
  if (!u) return false
  // 常见 TG 网页/图片 CDN（历史数据可能未写 source）
  if (u.includes('telesco.pe')) return true
  if (u.includes('telegram-cdn.org')) return true
  if (u.includes('cdn.telegram.org')) return true
  if (u.includes('telegram.org/file')) return true
  return false
}

/**
 * 按系统配置的返代模板拼接图片地址（豆瓣封面、TG 外链图等）。
 * 支持 `{url}`、以 `url=` 结尾、或自动追加 `?url=` / `&url=`。
 */
export function buildProxiedImageSrc(cover: string | undefined | null, template: string): string {
  const c = String(cover || '').trim()
  if (!c) return ''

  const tmpl = String(template || '').trim()
  if (!tmpl) return c

  const enc = encodeURIComponent(c)

  if (tmpl.includes('{url}')) return tmpl.replace(/\{url\}/g, enc)

  if (tmpl.endsWith('url=')) return tmpl + enc

  const idx = tmpl.indexOf('url=')
  if (idx >= 0) {
    const prefix = tmpl.slice(0, idx + 'url='.length)
    return prefix + enc
  }

  if (tmpl.includes('?')) {
    if (tmpl.endsWith('?') || tmpl.endsWith('&')) return tmpl + 'url=' + enc
    return tmpl + '&url=' + enc
  }
  return tmpl.replace(/\/?$/, '') + '?url=' + enc
}

File: dfan_netdisk_app/lib/utils/cover_proxy.dart
/// 是否与 Web 端 `shouldApplyTgCoverProxy` 一致
bool shouldApplyTgCoverProxy({
  required String source,
  required String externalId,
  required String coverRaw,
}) {
  if (source.trim().toLowerCase() == 'telegram') return true;
  final ext = externalId.trim().toLowerCase();
  if (ext.startsWith('tg:')) return true;
  final u = coverRaw.trim().toLowerCase();
  if (u.isEmpty) return false;
  if (u.contains('telesco.pe')) return true;
  if (u.contains('telegram-cdn.org')) return true;
  if (u.contains('cdn.telegram.org')) return true;
  if (u.contains('telegram.org/file')) return true;
  return false;
}

/// 与后台 / Web 端一致的图片返代拼接（如 `https://wsrv.nl/?url=`）。
String buildProxiedImageSrc(String cover, String template) {
  final c = cover.trim();
  if (c.isEmpty) return '';

  final tmpl = template.trim();
  if (tmpl.isEmpty) return c;

  final enc = Uri.encodeComponent(c);

  if (tmpl.contains('{url}')) {
    return tmpl.replaceAll('{url}', enc);
  }
  if (tmpl.endsWith('url=')) {
    return tmpl + enc;
  }
  final idx = tmpl.indexOf('url=');
  if (idx >= 0) {
    final prefix = tmpl.substring(0, idx + 'url='.length);
    return '$prefix$enc';
  }
  if (tmpl.contains('?')) {
    if (tmpl.endsWith('?') || tmpl.endsWith('&')) {
      return '${tmpl}url=$enc';
    }
    return '$tmpl&url=$enc';
  }
  var base = tmpl;
  if (base.endsWith('/')) {
    base = base.substring(0, base.length - 1);
  }
  return '$base?url=$enc';
}

/// 从 API Base（如 `http://host:8080/api/v1`）得到站点 origin，用于拼相对封面路径。
String apiOriginFromBase(String apiBaseUrl) {
  final u = Uri.parse(apiBaseUrl.trim());
  if (!u.hasScheme || u.host.isEmpty) return '';
  final port = u.hasPort ? ':${u.port}' : '';
  return '${u.scheme}://${u.host}$port';
}

/// 将封面字段解析为可请求的绝对 URL（不套返代）。
String resolveCoverToAbsoluteUrl(String cover, String apiBaseUrl) {
  final t = cover.trim();
  if (t.isEmpty) return '';

  final origin = apiOriginFromBase(apiBaseUrl);
  final scheme = Uri.tryParse(apiBaseUrl.trim())?.scheme ?? 'https';

  if (t.startsWith('//')) {
    return '$scheme:$t';
  }
  if (t.toLowerCase().startsWith('http://') || t.toLowerCase().startsWith('https://')) {
    return t;
  }
  if (origin.isEmpty) return t;
  return '$origin${t.startsWith('/') ? '' : '/'}$t';
}

/// TG 外链封面在配置返代后用于展示的最终 URL。
String resolveResourceCoverDisplayUrl({
  required String cover,
  required String source,
  String externalId = '',
  required String proxyTemplate,
  required String apiBaseUrl,
}) {
  final resolved = resolveCoverToAbsoluteUrl(cover, apiBaseUrl);
  if (resolved.isEmpty) return '';

  final raw = cover.trim();
  final isRemote = RegExp(r'^https?://', caseSensitive: false).hasMatch(raw) ||
      raw.startsWith('//');
  final tmpl = proxyTemplate.trim();
  if (tmpl.isNotEmpty &&
      isRemote &&
      shouldApplyTgCoverProxy(
        source: source,
        externalId: externalId,
        coverRaw: raw,
      )) {
    return buildProxiedImageSrc(resolved, tmpl);
  }
  return resolved;
}
